src/core/application/mod.rs

```diff
@@ -0,0 +1,179 @@
+//! The Application is the core component of crius.
+//! It manages the setup of the engine's sub-systems and
+//! runs the main loop.
+
+pub mod settings;
+
+use crate::core::application::settings::Settings;
+use crate::core::scene::{Context, Scene, SceneManager};
+use crate::prelude::Schedule;
+use legion::schedule::{Builder, Runnable, Schedulable};
+use legion::system::SystemBuilder;
+use legion::world::{Universe, World};
+use serde_yaml;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use winit::event::Event;
+use winit::event_loop::ControlFlow;
+use winit::window::WindowBuilder;
+
+const APPLICATION_SETTINGS_FILE_NAME: &str = "settings.yml";
+
+pub struct Application<E: 'static> {
+    universe: Universe,
+    world: World,
+    scene_manager: SceneManager<'static, E>,
+    schedule: Schedule,
+    settings: Settings,
+}
+
+impl<E> Application<E> {
+    pub fn run(self) {
+        let Application {
+            mut universe,
+            mut world,
+            mut scene_manager,
+            mut schedule,
+            settings,
+        } = self;
+
+        scene_manager.initialize(Context::new(&mut universe, &mut world));
+
+        let event_loop = winit::event_loop::EventLoop::with_user_event();
+
+        let mut window_builder = WindowBuilder::new()
+            .with_title(settings.window.title)
+            .with_resizable(settings.window.resizeable)
+            .with_maximized(settings.window.maximized)
+            .with_visible(settings.window.visible)
+            .with_transparent(settings.window.transparent)
+            .with_decorations(settings.window.decorations)
+            .with_always_on_top(settings.window.always_on_top);
+
+        if let Some(size) = settings.window.size {
+            window_builder = window_builder.with_inner_size(size);
+        }
+
+        if let Some(min) = settings.window.min_size {
+            window_builder = window_builder.with_max_inner_size(min);
+        }
+
+        if let Some(max) = settings.window.max_size {
+            window_builder = window_builder.with_max_inner_size(max);
+        }
```

Comment:
This looks like you can only have one of `size|min_size|max_size`, but there's nothing in the `yaml` file hinting it. I tried to search if there was a way in  `yaml` to describe that there are multiple choices and you can only pick one, but haven't found it.

Reply:
The documentation on this is not great but it seems you can optionally specify any, or all of these three.

In the code above I am checking if the Option<T> values deserialized from the yaml file are `Some` (they have values) and only then setting the value to the window builder.

If you do not specify any of them the library will set reasonable default based on the underlying windowing implementation of the OS.

Reply:
Oh, sorry, my mistake: I was looking at the left-hand-side, where `window_builder` was set, thinking it was overwritten, but it was already use on the right-hand-side, so it's previous value was used. This is fine :-)

Reply:
Yeah all these self re-assignments you see are not overwrites, but moves. Rust always moves unless you explicitly call .clone().

Because the `WindowBuilder` consumes `self` (its methods moves self instead of borrowing it) if you do not reassign to the other variable the builder will be moved into the scope of the if statement, the method will return `self` and that `self` temporary will be dropped at the end of the scope.

By reassigning the moved value I extend its lifetime to the outer scope instead of letting it die in the inner scope.